src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { Multiple = true });
+        }
+
+        public static void PushOperationWithContext(JSProxyContext knownContext)
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            // because this is called multiple times for each JSImport
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { CapturedContext = knownContext, Multiple = true });
+        }
+
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+
+            var current = AssertIsInteropThread();
+            stack.Add(new PendingOperation { CapturedContext = current });
+            return current;
+        }
+
+        public static void PopOperation()
+        {
+            var stack = OperationStack;
+            if (stack.Count < 1) Environment.FailFast("Unbalanced PopOperation");// there is no recovery
+            stack.RemoveAt(stack.Count - 1);
+        }
+
+        // this is here until we change the code generator and the API to Push/Pop the context
+        // we have no way how to Pop from the after marshaling the result
+        public static void SchedulePopOperation()
+        {
+            var stack = OperationStack;
+            var op = stack[stack.Count - 1];
+            if (op.Called) Environment.FailFast("SchedulePopOperation called twice");
+            op.Called = true;
+        }
+
+        public static void AssertOperationStack(int expected)
+        {
+            var stack = OperationStack;
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            var actual = stack.Count;
+            var multiple = stack.Count > 0 ? stack[stack.Count - 1].Multiple : false;
+            var called = stack.Count > 0 ? stack[stack.Count - 1].Called : false;
+            if (actual != expected) Environment.FailFast($"Unexpected OperationStack size expected: {expected} actual: {actual} called:{called} multiple:{multiple}");
+        }
+
+        // TODO: sort generated ToJS() calls to make the capture context before we need to use it
+        public static void CaptureContextFromParameter(JSProxyContext parameterContext)
+        {
+            var stack = OperationStack;
+            if (stack.Count < 1) Environment.FailFast("CaptureContextFromParameter could be only used during pending operation.");
```

Comment:
Should we automatically perform scheduled pops here first? It seems like if we're capturing contexts from parameters then we are in the process of marshaling an argument list, so there shouldn't be any on the stack. Or can context be captured during return value marshaling?

Reply:
When we push new frame, we pop any scheduled already. We can assert, that current is not scheduled.
No, context can't be captured during return value marshaling.